src/lib/rpc/bestRpcTracker.ts

```diff
@@ -0,0 +1,343 @@
+  const controller = new AbortController();
+
+  const startTime = Date.now();
+
+  return await Promise.race([
+    sleep(PROBE_FAIL_TIMEOUT).then(() => {
+      controller.abort();
+      throw new Error("Probe timeout");
+    }),
+    (async function callRpc() {
+      const dataStoreAddress = getContract(chainId, "DataStore");
+      const multicallAddress = getContract(chainId, "Multicall");
+      const dataStore = getDataStoreContract(chainId, provider);
+      const multicall = getMulticallContract(chainId, provider);
+
+      const probeMarketAddress = PROBE_SAMPLE_MARKET[chainId];
+      const probeFieldKey = HASHED_MARKET_CONFIG_KEYS[chainId]?.[probeMarketAddress]?.[PROBE_SAMPLE_FIELD];
+
+      let blockNumber: number | null = null;
+      let isSuccess = false;
+
+      if (!dataStoreAddress || !multicallAddress || !probeMarketAddress || !probeFieldKey || !dataStore || !multicall) {
+        throw new Error("Failed to get RPC probe request data");
+      } else {
+        const dataStoreData = dataStore.interface.encodeFunctionData("getUint", [probeFieldKey]);
+        const multicallData = multicall.interface.encodeFunctionData("blockAndAggregate", [
+          [{ target: dataStoreAddress, callData: dataStoreData }],
+        ]);
+
+        const body = {
+          jsonrpc: "2.0",
+          id: 1,
+          method: "eth_call",
+          params: [
+            {
+              to: multicallAddress,
+              data: multicallData,
+            },
+            "latest",
+          ],
+        };
+
+        try {
+          const response = await fetch(providerUrl, {
+            method: "POST",
+            headers: {
+              "Content-Type": "application/json",
+            },
+            body: JSON.stringify(body),
+            signal: controller.signal,
+          });
+
+          if (!response.ok) {
+            throw new Error(`Network response was not ok: ${response.statusText}`);
+          }
+
+          const { result } = await response.json();
+
+          if (result) {
+            const multicallResult = multicall.interface.decodeFunctionResult("blockAndAggregate", result);
+            const [sampleFieldValue] = dataStore.interface.decodeFunctionResult(
+              "getUint",
+              multicallResult.returnData[0].returnData
+            );
+
+            blockNumber = Number(multicallResult.blockNumber);
+            isSuccess = sampleFieldValue && sampleFieldValue > 0;
+          }
+        } catch (error) {
+          if (error.name !== "AbortError") {
+            throw error;
+          }
+        }
+      }
+
+      return {
+        url: providerUrl,
+        responseTime: Date.now() - startTime,
+        blockNumber,
+        timestamp: new Date(),
+        isSuccess,
+      };
+    })(),
+  ]).catch(() => {
+    return {
+      url: providerUrl,
+      responseTime: Date.now() - startTime,
```

Comment:
btw is there a reason to return response time for failed requests? i guess time of failed response should not be considered anyway?